meson: allow unsetting the --buildtype flag

Twey opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Meson allows setting properties in many ways.  In our case we have the buildtype set in the Meson project options file, and we would like to be able to ask the nixpkgs setup hook not to override it. 

This PR does not change the default behaviour (which is to specify `--buildtype=plain`), but allows the user to not pass the `--buildtype` flag at all by explicitly setting `mesonBuildType = ""`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Twey committed:
meson: allow unsetting the --buildtype flag

jtojnar commented:
What is the motivation behind this? How would this differ from just seeing it to the build type of your choice? 

What is `@Q`?

Also this should go to staging branch.

Twey commented:
`@Q` escapes a string for safe interpolation into a bash command.

I'm confused by the ‘not a best practice’ comment.  Surely it's actually a sensible default to respect the options the project specifies it should be built with, and only override them if explicitly requested?  Maybe I'm missing something philosophical here.

Twey commented:
Should I be directing it into `staging` or `staging-next`?

jtojnar commented:
Oh, it is documented as `parameter@operator` in the [bash manual](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Parameter-Expansion).

It does not appear to prevent the `mesonBuildType` from splitting, thought:

```bash
mesonBuildType="test foo"
mesonFlags="--prefix=$prefix $mesonFlags"

mesonBuildType="${mesonBuildType-plain}"
mesonBuildTypeFlag="${mesonBuildType:+--buildtype=${mesonBuildType@Q} }"
mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonBuildTypeFlag$mesonFlags"
python3 -c 'import sys; print(sys.argv)' build $mesonFlags "${mesonFlagsArray[@]}"
```

produces `['-c', 'build', "--buildtype='test", "foo'", '--prefix=']`.

jtojnar commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -21,7 +21,9 @@ mesonConfigurePhase() {
         $mesonFlags"
 
-    mesonFlags="${crossMesonFlags+$crossMesonFlags }--buildtype=${mesonBuildType:-plain} $mesonFlags"
+    mesonBuildType="${mesonBuildType-plain}"
+    mesonBuildTypeFlag="${mesonBuildType:+--buildtype=${mesonBuildType@Q} }"
+    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonBuildTypeFlag$mesonFlags"
```
I do not really like the bash shell parameter expansion as I can never remember which combination of symbols does what and it is hard to search for. Using the `if` statement is clearer IMO:

```suggestion
    # By default, we use plain build type. You can override it using mesonBuildType
    # variable or disable passing the flag by setting the variable to empty string.
    mesonBuildType="${mesonBuildType-plain}"
    if [[ -n $mesonBuildType ]]; then
        mesonFlags="--buildtype=${mesonBuildType} $mesonFlags"
    fi
    
    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonFlags"
```

The suggestion also removes the ineffectual `@Q` (we should really switch to using an array for all the flags) and adds an explanation comment since the empty string behaviour is easy to miss.

jtojnar replied:
Actually, having a special empty value might be confusing as well. I would prefer just using an explicit switch:

```suggestion
    if [[ -z "${dontAddMesonBuildType-}" ]]; then
        mesonFlags="--buildtype=${mesonBuildType:-plain} $mesonFlags"
    fi

    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonFlags"
```

Twey replied:
@jtojnar What makes you think the `@Q` is ineffectual?  I think it is required to make `mesonBuildType` values containing spaces work properly (admittedly it happens there are no valid such, but at least it will make the error behaviour a little less confusing).
Indeed it would ideally be an array, but I'm not keen on changing that interface in this PR about removing a flag.

On the use of parameter expansion features — I think it's quite idiomatic bash, but at any rate I'd rather follow the existing style rather than refactoring (and this refactor uses the `-` parameter expansion form anyway).

jtojnar replied:
Ineffectual I referred to https://github.com/NixOS/nixpkgs/pull/82264#issuecomment-603274127 – `$mesonFlags` is [passed to meson](https://github.com/Twey/nixpkgs/blob/1a773610af63351a6bdf91c0427667b831c349f1/pkgs/development/tools/build-managers/meson/setup-hook.sh#L30) so `IFS` splitting occurs and quoting does not help.

Personally, I prefer slightly more verbose `if` expressions since they are much more understandable to people without deep Bash background (and the variable expansion forms are hardly googleable) but I am not categorically opposed to using those Bash features. But I definitely think using an empty string value is too obscure and would vote for `dontAddMesonBuildType` as mentioned [above](https://github.com/NixOS/nixpkgs/pull/82264/files#r397209274).

Twey replied:
I'm not super keen on the flag because it adds nonsensical cases where the `mesonBuildType` is set but also `dontAddMesonBuildType`.  Ideally I would have had the undefined case not set the build type, and let the user or `mkDerivation` decide when they want to override it, but I think that would change the behaviour of too many packages now, and I agree that the special empty string behaviour is perhaps a little surprising if not documented. 

I've accepted the suggestion: I've a mild bias towards preserving the status quo, but better safe than sorry on the readability aspect. :)

jtojnar commented:
> Should I be directing it into `staging` or `staging-next`?

`staging` is the place where mass-rebuilds are accumulated so it should go there. Once in a while, commits from `staging` are merged to `staging-next` and once the CI rebuilds the world on there, it is merged to `master`.

Twey committed:
meson: add dontAddMesonBuildType flag

Add a `dontAddMesonBuildType` flag that disables adding `--buildtype=plain` to the Meson command.

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>

jtojnar commented:
Could you also please document it?

Adding something like `Can be disabled by passing <code>dontAddMesonBuildType = true;</code>.` to https://github.com/NixOS/nixpkgs/blob/3f61850c410602d1f80f947ab78e38335f1a215d/doc/stdenv/stdenv.xml#L2140-L2141 should work.
